function to parse an interval in a SAMSequenceDictionary + tests

lindenb opened the pull request:
Added a small function to parse an interval + test in  SAMSequenceDictionary .

```java
 public Interval parseInterval(final String region) 
```

The function checks that the contig exist in the dict, and clip the positions if pos<1 or pos>contig.length()

Various syntaxes are supported:

     *     CHROM -> CHROM:1-seq.length()
     *     CHROM:POS ->  CHROM:POS-POS
     *     CHROM:POS+NUM -> CHROM:{POS-NUM}-{POS+NUM}
     *     CHROM:BEGIN- ->  CHROM:BEGIN-seq.length()
     *     CHROM:BEGIN-END ->  CHROM:BEGIN-END

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+     * @param numStr the coordinate
+     * @return the coordinate as integer
+     */
+    private static int parseCoordinate(final String numStr)
```
is SamDictionary the right place for this? isn't this a way to define an Interval? seems like Interval is a more natural place for it. What do you think?

yfarjoun replied:
Nevermind. You are using the dictionary itself for the contig order. Ignore.

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+        if(num.endsWith("bp"))
+            {
+            factor = 1L;
+            num = num.substring(0,num.length()-2).trim();
```
refactor out this line please, and fix the alignment:

```code
if␣(num.endsWith("bp"))␣{
            factor = 1L;
            num = num.substring(0,␣num.length()␣-␣2).trim();
} else { 
``` 

etc.

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+            }
+        else if(num.endsWith("gb"))
+            {
+            factor = 1000000000L;//1E9
```
use underscores for easy reading:
```code
factor = 1_000_000_000L; / 1E9
```

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+            factor = 1000000000L;//1E9
+            num = num.substring(0,num.length()-2).trim();
+            }
+        
```
should you throw an exception if there substring doesn't match any of the things you attempted?

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+        try {
+            long pos = Long.parseLong(num);
+            pos *= factor;
+            if(pos < 1L || pos > Integer.MAX_VALUE)  throw new IllegalArgumentException("Cannot convert "+numStr+" to a positive integer");
```
double space and missing spaces around `+` signs

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+            if(pos < 1L || pos > Integer.MAX_VALUE)  throw new IllegalArgumentException("Cannot convert "+numStr+" to a positive integer");
+            return (int)pos;
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Cannot convert "+numStr+" to a coordinate");
```
spaces

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+    
+    /** 
+     * parses the interval String and returns a htsjdk.samtools.util.Interval
+     * The contig/chromosome must exist in the dictionary. out-of-bound index <code>!(1<= index <= sequence.length())</code> will be clipped 
```
above zero throws an exception, doesn't clip. please correct comment.

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+    /** 
+     * parses the interval String and returns a htsjdk.samtools.util.Interval
+     * The contig/chromosome must exist in the dictionary. out-of-bound index <code>!(1<= index <= sequence.length())</code> will be clipped 
+     * coordinate suffixes like 'bp,Mb,Gb' will be interpreted.
```
...as 1,1000,and 1000000 bases

yfarjoun commented on src/tests/java/htsjdk/samtools/SAMSequenceDictionaryTest.java:
```diff
@@ -39,6 +41,55 @@
+        final SAMSequenceDictionary dict = new SAMSequenceDictionary(
+                Arrays.asList(ssr1));
+        Assert.assertEquals(dict.size(), 1);
+        Assert.assertEquals(
```
please add test for "1:0"

yfarjoun replied:
please use a @DataProvider

yfarjoun replied:
please add tests that have spaces in them (not only in the begining and end)

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+     * Position=0 will be set to 1.
+     * A Position larger than the contig size will be set to the contig size
+     * 
+     * Valid syntax -> convertion :
```
conversion

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+                    record.getSequenceLength()
+                    );
+        }
+        if(colon == 0)  throw new IllegalArgumentException("Empty chromosome in "+region);
```
lots of spaces missing from here and down. esp. around `+` signs.

yfarjoun replied:
and after commas

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+     * coordinate suffixes like 'bp,Mb,Gb' will be interpreted.
+     * Comma in the coordinate will be removed.
+     * left/right white spaces will be trimmed.
+     * Position=0 will be set to 1.
```
comment disagrees with code.

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+            final int pos = parseCoordinate(region.substring(colon+1,plus));
+            final int extend = parseCoordinate(region.substring(plus+1));
+            if( extend < 0)  throw new IllegalArgumentException("Extend cannot be negative in "+region);
+            if( pos - extend  > record.getSequenceLength() || pos+extend  <1 ) throw new IllegalArgumentException("In "+region +", position="+pos+"+/-"+extend+" out of bounds 1<=pos<="+record.getSequenceLength());
```
code disagrees with java doc. are you clipping or throwing an exception?

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+        if(hyphen == -1 && plus > colon) {
+            final int pos = parseCoordinate(region.substring(colon+1,plus));
+            final int extend = parseCoordinate(region.substring(plus+1));
+            if( extend < 0)  throw new IllegalArgumentException("Extend cannot be negative in "+region);
```
parseCoordinate would never return a negative number...

yfarjoun commented on src/tests/java/htsjdk/samtools/SAMSequenceDictionaryTest.java:
```diff
@@ -39,6 +41,55 @@
+                new Interval("1", 10, contigLength)
+               );
+    }
+    
```
could you please add some test that expectedly throw exceptions due to formatting of input string?

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+           }
+       //CHROM:START-END
+       final int chromEnd = parseCoordinate(region.substring(hyphen+1));
+       if(chromEnd < chromStart)  throw new IllegalArgumentException("In "+region +", start>end");
```
Interval actually supports the end being 1 less than the start to indicate an empty interval between the two bases:

1:10-9 can point to the place of an insertion between the 9th and 10th bases. Would you like to allow for that in the parsing?

yfarjoun commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -215,9 +217,128 @@ public SAMSequenceRecord addSequenceAlias(final String originalName,
+        final int colon = region.indexOf(':');
+        // no colon, chromosome only
+        if(colon == -1) {
+            final SAMSequenceRecord record = getSequence(region);
```
There's a lot of parsing going on here. Would it be possible to use a regex to capture the parsing? 
something like this:
```code
^(.*?)((:)(.*?)(([-+])?(.*?))?)?$
```

Then the first group would be the CHROM
if there is a second group, the fourth would be POS1 the sixth (if there) would be the + or -  and the seventh (if there) would the POS2, then you can separate the parsing from the logic of constructing the interval, which currently is combined in a complicated way.

I'm not saying it has to be my regex...that is just an example...but I think it would be good to separate the parsing of the string and the logic of creating the Interval.

lbergelson replied:
I'd definitely vote against having a regex like that.  Separating the parsing from the interpretation is a good idea though.

yfarjoun commented on src/tests/java/htsjdk/samtools/SAMSequenceDictionaryTest.java:
```diff
@@ -39,6 +41,55 @@
+               );
+        Assert.assertEquals(
+                dict.parseInterval("1:10-10kb"),
+                new Interval("1", 10, contigLength)
```
mb is not tested 
kb isn't tested to return the correct interval (only clipped)

yfarjoun commented:
review done.

lindenb commented:
@yfarjoun  One month later I'm back to this pull request.

   * As you suggested, I've moved everything to InrervalUtil (rather than Interval or SamSequenceDictionary )
   * I've added positive and negative tests, using DataProvider 
   * updated the javadoc, messages, etc...
  
thanks.

droazen commented:
@yfarjoun Is this one ok to merge? Has @lindenb has addressed your original comments to your satisfaction?
